macros/tables/sat.sql

```diff
@@ -11,41 +11,45 @@
 {{ dbtvault.prepend_generated_by() }}
 
 WITH source_data AS (
-    SELECT *
-    FROM {{ ref(source_model) }}
+    {%- if model.config.materialized == 'vault_insert_by_rank' %}
+    SELECT {{ dbtvault.prefix(source_cols_with_rank, 'a', alias_target='source') }}
+    {%- else %}
+    SELECT {{ dbtvault.prefix(source_cols, 'a', alias_target='source') }}
+    {%- endif %}
+    FROM {{ ref(source_model) }} AS a
+    WHERE {{ dbtvault.multikey(src_pk, prefix='a', condition='IS NOT NULL') }}
     {%- if model.config.materialized == 'vault_insert_by_period' %}
     WHERE __PERIOD_FILTER__
     {% endif %}
 ),
 
 {% if dbtvault.is_vault_insert_by_period() or is_incremental() -%}
 
-latest_sat_records_in_source AS (
+latest_records AS (
+
     SELECT {{ dbtvault.prefix(source_cols, 'a', alias_target='target') }}
-    FROM {{ this }} as a
-    JOIN source_data as b
-    ON a.{{ src_pk }} = b.{{ src_pk }}
-    WHERE 1 = 1
-    QUALIFY
-        1 = RANK() OVER (
-            PARTITION BY {{ dbtvault.prefix([src_pk], 'a') }}
-            ORDER BY {{ dbtvault.prefix([src_ldts], 'a') }} DESC
-        )
+    FROM (
+        SELECT {{ dbtvault.prefix(source_cols, 'current_records', alias_target='target') }},
+        RANK() OVER (
+        PARTITION BY {{ dbtvault.prefix([src_pk], 'current_records') }}
+        ORDER BY {{ dbtvault.prefix([src_ldts], 'current_records') }} DESC
+        ) AS rank
+        FROM {{ this }} AS current_records
+        ) a
+    WHERE a.rank = 1
 ),
 
 {% endif -%}
 
 records_to_insert AS (
-    SELECT DISTINCT {{ dbtvault.alias_all(source_cols, 'c') }}
-    FROM source_data AS c
-    {% if dbtvault.is_vault_insert_by_period() or is_incremental() -%}
-    LEFT JOIN
-        latest_sat_records_in_source
-    ON
-        {{ dbtvault.prefix([src_hashdiff], 'latest_sat_records_in_source', alias_target='target') }} = {{ dbtvault.prefix([src_hashdiff], 'c') }}
+    SELECT DISTINCT {{ dbtvault.alias_all(source_cols, 'stage') }}
+    FROM source_data AS stage
+    {%- if dbtvault.is_any_incremental() %}
+    LEFT JOIN latest_records
+    ON {{ dbtvault.prefix([src_pk], 'latest_records', alias_target='target') }} = {{ dbtvault.prefix([src_pk], 'stage') }}
     WHERE
-        {{ dbtvault.prefix([src_hashdiff], 'latest_sat_records_in_source', alias_target='target') }} IS NULL
-    {% endif %}
+        {{ dbtvault.prefix([src_hashdiff], 'latest_records', alias_target='target') }} IS NULL        
```

Comment:
I don't see how this could be the right behaviour.

take for example a source where all the values have changed compared to the latest records in the sat.

As it stands in this PR, you'd **source LEFT JOIN sat_latest** on the PK.

Since all the PKs match there is not gonna be any rows where the `sat_latest` hashdiff is NULL.

This is essentially only gonna insert new records, but won't account for cases where the record has changed